apis/memorystore/v1alpha1/instance_types.go

```diff
@@ -0,0 +1,195 @@
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MemorystoreInstanceSpec defines the desired state of MemorystoreInstance
+// +kcc:proto=google.cloud.memorystore.v1beta.Instance
+type MemorystoreInstanceSpec struct {
+	// Optional. Immutable. The MemorystoreInstance name. If not given, the metadata.name will be used.
+	ResourceID *string `json:"resourceID,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Parent field is immutable"
+	// Immutable.
+	// +required
+	Parent `json:",inline"`
+
+	// Optional. Number of replica nodes per shard. If omitted the default is 0
+	//  replicas.
+	ReplicaCount *int32 `json:"replicaCount,omitempty"`
+
+	// Optional. Immutable. Authorization mode of the instance.
+	AuthorizationMode *string `json:"authorizationMode,omitempty"`
+
+	// Optional. Immutable. In-transit encryption mode of the instance.
+	TransitEncryptionMode *string `json:"transitEncryptionMode,omitempty"`
+
+	// Optional. Number of shards for the instance.
+	ShardCount *int32 `json:"shardCount,omitempty"`
+
+	// Optional. Immutable. Machine type for individual nodes of the instance.
+	NodeType *string `json:"nodeType,omitempty"`
+
+	// Optional. Persistence configuration of the instance.
+	PersistenceConfig *PersistenceConfig `json:"persistenceConfig,omitempty"`
+
+	// Optional. Immutable. Engine version of the instance.
+	EngineVersion *string `json:"engineVersion,omitempty"`
+
+	// Optional. User-provided engine configurations for the instance.
+	EngineConfigs map[string]string `json:"engineConfigs,omitempty"`
+
+	// Optional. Immutable. Zone distribution configuration of the instance for
+	//  node allocatiteon.
+	ZoneDistributionConfig *ZoneDistributionConfig `json:"zoneDistributionConfig,omitempty"`
+
+	// Optional. If set to true deletion of the instance will fail.
+	DeletionProtectionEnabled *bool `json:"deletionProtectionEnabled,omitempty"`
+
+	// Required. Immutable. User inputs for the auto-created PSC connections.
+	PscAutoConnectionsSpec []PscAutoConnectionSpec `json:"pscAutoConnections,omitempty"`
+}
+
+// kcc specific struct to separate input and output fiels in
+// google.cloud.memorystore.v1beta.PscAutoConnection
+type PscAutoConnectionSpec struct {
+
+	// Required. The consumer project_id where PSC connections are established.
+	//  This should be the same project_id that the cluster is being created in.
+	// +required
+	ProjectRef *refs.ProjectRef `json:"projectRef"`
+
+	// Required. The network where the PSC endpoints are created, in the form of
+	//  projects/{project_id}/global/networks/{network_id}.
+	// +required
+	NetworkRef *refs.ComputeNetworkRef `json:"networkRef,omitempty"`
+}
+
+type Parent struct {
+	// +required
+	ProjectRef *refs.ProjectRef `json:"projectRef"`
+
+	// +required
+	Location string `json:"location"`
```

Comment:
Is location immutable?

Reply:
Yes it is, I thought since I marked parent as immutable, it applies to underlying fields too. Do I need additional annotations?

Reply:
Updated it as per the example https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/cc54c3b56588103398a95a28b7e35f10071587b7/apis/bigquerydatatransfer/v1alpha1/config_types.go#L115-L123